honeybee_radiance/state.py

```diff
@@ -0,0 +1,744 @@
+from .lib.modifiers import white_glow
+
+from honeybee.shade import Shade
+from ladybug_geometry.geometry3d.face import Face3D
+
+import math
+
+
+class _RadianceState(object):
+    """Object representing a single state for a dynamic object.
+
+    Args:
+        modifier: A Honeybee Radiance Modifier object to be applied to this state's
+            parent in this state. This can be used to swap out aperture BSDFs
+            in 3-phase studies, change the modifier of a shade object representing
+            the ground to account for snow reflectance, etc. If None, it will
+            be the parent's default modifier.
+        shades: An optional array of honeybee-core Shade objects to be included
+            with this state. These shades must not have a parent Room, Face or
+            Aperture and they cannot already be assigned to another state.
+            Also, the Shade must not have a dynamic_group_identifier of its own.
+
+    Properties:
+        * modifier
+        * shades
+        * modifier_direct
+        * parent
+        * has_parent
+    """
+
+    __slots__ = ('_modifier', '_shades', '_modifier_direct', '_parent')
+
+    def __init__(self, modifier=None, shades=None):
+        """Initialize RadianceState."""
+        self._parent = None  # will be set when the state is assigned
+        self.modifier = modifier
+        self.shades = shades
+        self._modifier_direct = None
+
+    @property
+    def modifier(self):
+        """Get or set the modifier to be applied to the parent in this state.
+        """
+        if not self._modifier and self.has_parent:  # use the parent's default
+            return self.parent.properties.radiance.modifier
+        return self._modifier
+
+    @modifier.setter
+    def modifier(self, value):
+        if value is not None:
+            assert isinstance(value, Modifier), \
+                'Expected Modifier for RadianceState. Got {}'.format(type(value))
+            value.lock()  # lock editing in case modifier has multiple references
+        self._modifier = value
+
+    @property
+    def shades(self):
+        """Get or set an array of honeybee-core shades to be included with this state.
+    
+        These shades must not have a parent Room, Face or Aperture and they cannot
+        already be assigned to another RadianceState. Also, the Shade must not have
+        a dynamic_group_identifier of its own, meaning that a dynamic shade cannot
+        be assigned to another dynamic object.
+        """
+        return tuple(self._shades)
+
+    @shades.setter
+    def shades(self, value):
+        if value is not None:
+            try:
+                self._shades = [self._check_shade(sh) for sh in value]
+            except (ValueError, TypeError):
+                raise TypeError('RadianceState shades must be an iterable. ' \
+                    'Got  {}.'.format(type(value)))
+        else:
+            self._shades = []
+
+    @property
+    def modifier_direct(self):
```

Comment:
> I very much like the use of modifier_direct instead of modifier_blk. It is more explicit. Are we using modifier_blk only for isolation studies now?
Yes.  for dynamic apertures, `modifier_blk` is what is used in isolation studies while the `modifier_direct` is what is used in direct studies.  For dynamic shades, the `modifier_blk` is only used for direct studies if the shade is opaque. Otherwise, it will use the default modifier in direct studies of non-opaque shades. The idea behind this property here is that the user can always override the `modifier_direct` on a particular state if they want.

It might be worthwhile for you to just review the dynamic folder of the radiance-folder-structure since this PR is really trying to reproduce the conditions that are in that repo. Generally, speaking, if there is something wrong or odd in that repo, then the same issue is going to show up here.

Reply:
Sounds good. Now that we have the folder structure we can start thinking about improvements. Thanks.